clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -1060,236 +1061,355 @@ CodeGenFunction::evaluateOrEmitBuiltinObjectSize(const Expr *E, unsigned Type,
-    // what GCC does for consistency's sake.
+  const FieldDecl *CountFD = FlexibleArrayMemberFD->findCountedByField();
+  if (!CountFD)
+    // Can't find the field referenced by the "counted_by" attribute.
     return nullptr;
 
-  // Build a load of the counted_by field.
-  bool IsSigned = CountedByFD->getType()->isSignedIntegerType();
-  Value *CountedByInst = EmitLoadOfCountedByField(Base, FAMDecl, CountedByFD);
-  if (!CountedByInst)
-    return getDefaultBuiltinObjectSizeResult(Type, ResType);
-
-  CountedByInst = Builder.CreateIntCast(CountedByInst, ResType, IsSigned);
+  const Expr *Idx = nullptr;
+  if (Visitor.ASE) {
+    Idx = Visitor.ASE->getIdx();
 
-  // Build a load of the index and subtract it from the count.
-  Value *IdxInst = nullptr;
-  if (Idx) {
-    if (Idx->HasSideEffects(getContext()))
+    if (Idx->HasSideEffects(Ctx))
       // We can't have side-effects.
       return getDefaultBuiltinObjectSizeResult(Type, ResType);
 
+    if (const auto *IL = dyn_cast<IntegerLiteral>(Idx)) {
+      int64_t Val = IL->getValue().getSExtValue();
+      if (Val < 0)
+        return getDefaultBuiltinObjectSizeResult(Type, ResType);
+
+      // The index is 0, so we don't need to take it into account.
+      if (Val == 0)
+        Idx = nullptr;
+    }
+  }
+
+  // Calculate the flexible array member's object size using these formulae
+  // (note: if the calculation is negative, we return 0.):
+  //
+  //      struct p;
+  //      struct s {
+  //          /* ... */
+  //          int count;
+  //          struct p *array[] __attribute__((counted_by(count)));
+  //      };
+  //
+  // 1) 'ptr->array':
+  //
+  //    size_t count = (size_t) ptr->count;
+  //
+  //    size_t flexible_array_member_base_size = sizeof (*ptr->array);
+  //    size_t flexible_array_member_size =
+  //            count * flexible_array_member_base_size;
+  //
+  //    if (flexible_array_member_size < 0)
+  //        return 0;
+  //    return flexible_array_member_size;
+  //
+  // 2) '&ptr->array[idx]':
+  //
+  //    size_t count = (size_t) ptr->count;
+  //    size_t index = (size_t) idx;
+  //
+  //    size_t flexible_array_member_base_size = sizeof (*ptr->array);
+  //    size_t flexible_array_member_size =
+  //            count * flexible_array_member_base_size;
+  //
+  //    size_t index_size = index * flexible_array_member_base_size;
+  //
+  //    if (flexible_array_member_size < 0 || index < 0)
+  //        return 0;
+  //    return flexible_array_member_size - index_size;
+  //
+  // 3) '&ptr->field':
+  //
+  //    size_t count = (size_t) ptr->count;
+  //    size_t sizeof_struct = sizeof (struct s);
+  //
+  //    size_t flexible_array_member_base_size = sizeof (*ptr->array);
+  //    size_t flexible_array_member_size =
+  //            count * flexible_array_member_base_size;
+  //
+  //    size_t field_offset = offsetof (struct s, field);
+  //    size_t offset_diff = struct_size - field_offset;
```

Comment:
```suggestion
  //    size_t offset_diff = sizeof_struct - field_offset;
```

Reading through the comments, trying to understand more precisely the 4 different cases. Seems in this comment that `offset_diff` is unused?

Reply:
Done.